File: elixir/math/matrix.exs
# Multiply two fixed 3x3 integer matrices, reaching the right one's columns with zip.
# Run: elixir matrix.exs

defmodule Matrix do
  def multiply(left, right) do
    columns = right |> Enum.zip() |> Enum.map(&Tuple.to_list/1)

    for row <- left do
      for column <- columns do
        row |> Enum.zip(column) |> Enum.map(fn {x, y} -> x * y end) |> Enum.sum()
      end
    end
  end

  def determinant([[a, b, c], [d, e, f], [g, h, i]]) do
    a * (e * i - f * h) - b * (d * i - f * g) + c * (d * h - e * g)
  end
end

left = [[2, -1, 0], [1, 3, 4], [0, 5, -2]]
right = [[1, 0, 2], [-3, 1, 1], [4, 2, 0]]
product = Matrix.multiply(left, right)

Enum.each(product, fn row -> IO.puts(Enum.join(row, " ")) end)
IO.puts(Matrix.determinant(product))

File: scala/math/Matrix.scala
// Multiply two fixed 3x3 integer matrices, reaching the right one's columns with transpose.
// Run: scala-cli run Matrix.scala

val left = Vector(Vector(2, -1, 0), Vector(1, 3, 4), Vector(0, 5, -2))
val right = Vector(Vector(1, 0, 2), Vector(-3, 1, 1), Vector(4, 2, 0))

def multiply(a: Vector[Vector[Int]], b: Vector[Vector[Int]]): Vector[Vector[Int]] =
  val columns = b.transpose
  a.map(row => columns.map(column => row.zip(column).map((x, y) => x * y).sum))

def determinant(m: Vector[Vector[Int]]): Int =
  m(0)(0) * (m(1)(1) * m(2)(2) - m(1)(2) * m(2)(1)) -
    m(0)(1) * (m(1)(0) * m(2)(2) - m(1)(2) * m(2)(0)) +
    m(0)(2) * (m(1)(0) * m(2)(1) - m(1)(1) * m(2)(0))

@main def matrixMain(): Unit =
  val product = multiply(left, right)
  product.foreach(row => println(row.mkString(" ")))
  println(determinant(product))
